components/sup/src/manager/mod.rs

```diff
@@ -677,8 +681,17 @@ impl Manager {
         if self.http_disable {
             info!("http-gateway disabled");
         } else {
+            // Before starting up the HTTP server, let's double check that the listen address we
+            // have is available for binding. We want to check this up front before trying to bind
+            // because the binding happens in a background thread, and if it fails, the supervisor
+            // will continue running but the user may not notice that the HTTP server is dead in
+            // the water.
+            if TcpStream::connect(&http_listen_addr.to_string()).is_ok() {
```

Comment:
This should be `is_err`, shouldn't it?

Strangely, I see `TcpStream::connect` return `Err(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })` even when the bind is eventually successful (confirmed with a `curl`), so I think there's something wrong here.

Also, I think this approach isn't quite right since I think you'd need a `shutdown` before the `bind`. Unless I'm missing it, `TcpStream` doesn't seem to have any `drop` behavior.

Rather than doing all that, can we try the other approach? There are additional reasons a `bind` could fail other than just the address being in use, and trying to check for all possible errors in `http_gateway::Server::run` on the caller side seems sort of awkward to me. I'm happy to work on the mechanics of getting the data in and out of the thread together if you'd like.

Reply:
No, `is_ok` is correct because if there's something else already listening on the port we want to bind on, then this connection will succeed, which means the bind will fail.  I tested this with another service running on the HTTP port before starting the supervisor up and it properly fails.

`TcpStream`'s drop behavior is that the connection is dropped when the variable holding onto it goes out of scope, but I don't think that's necessary in this case because either the connection succeeds, in which case we immediately return an `Err` and the supervisor exits, or the connection fails, in which case it doesn't matter, because the bind in the other thread succeeds.  I tested this manually in a few different scenarios and everything behaves the way you would expect.

I can definitely try the other approach, but this seemed a lot simpler and more straightforward to me, which is why I did it first.

Reply:
Oh! I see what you're doing here. I was thinking about it backwards, like you were trying the operation that the bind would do, not that you were simulating the client. That makes more sense now.

And you're right that `TcpStream` shuts down the connection on drop. I had looked for that in the docs, but missed it somehow.

Nonetheless, there are other reasons that the `bind` could fail. For example, it's still possible to get a panic in the http-gateway thread with the code as is and only mildly contrived input:

```
➤ env TESTING_FS_ROOT=(mktemp -d)  HAB_SUP_BINARY=(pwd)/target/debug/hab-sup target/debug/hab sup run --listen-http 0.0.0.0:666
…
hab-sup(MR): core/hab-sup (core/hab-sup/0.66.0-dev)
hab-sup(MR): Supervisor Member-ID 8881e02feceb4d04a6a4e6f8ad4c4572
hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
hab-sup(MR): Starting ctl-gateway on 127.0.0.1:9632
hab-sup(MR): Testing connecting http-gateway on 0.0.0.0:666
hab-sup(MR): connect_result: Err(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" })
hab-sup(MR): Testing connecting http-gateway on 0.0.0.0:666 passed
hab-sup(MR): Starting http-gateway on 0.0.0.0:666
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', libcore/result.rs:945:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:475
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:390
   6: rust_begin_unwind
             at libstd/panicking.rs:325
   7: core::panicking::panic_fmt
             at libcore/panicking.rs:77
   8: core::result::unwrap_failed
             at /checkout/src/libcore/macros.rs:26
   9: <core::result::Result<T, E>>::unwrap
             at /checkout/src/libcore/result.rs:782
  10: habitat_sup::http_gateway::Server::run::{{closure}}
             at components/sup/src/http_gateway.rs:149
```

And if there were implementation changes that added new failure modes on the callee side of `http_gateway::Server::run`, I think it's a cleaner abstraction for the caller not to have to worry about those.